keras_lr_finder/lr_finder.py

```diff
@@ -52,38 +51,33 @@ def find(self, x_train, y_train, start_lr, end_lr, batch_size=64, epochs=1):
 
         callback = LambdaCallback(on_batch_end=lambda batch, logs: self.on_batch_end(batch, logs))
 
-        self.model.fit(x_train, y_train,
-                        batch_size=batch_size, epochs=epochs,
-                        callbacks=[callback])
+        # Calling fit or fit_generator
+        fit_fn(self.model, callback)
 
         # Restore the weights to the state before model fitting
         self.model.load_weights('tmp.h5')
 
         # Restore the original learning rate
         K.set_value(self.model.optimizer.lr, original_lr)
 
-    def find_generator(self, generator, start_lr, end_lr, batch_size=64, epochs=1):
-        num_batches = epochs * len(generator)
-        self.lr_mult = (float(end_lr) / float(start_lr)) ** (float(1) / float(num_batches))
-
-        # Save weights into a file
-        self.model.save_weights('tmp.h5')
-
-        # Remember the original learning rate
-        original_lr = K.get_value(self.model.optimizer.lr)
-
-        # Set the initial learning rate
-        K.set_value(self.model.optimizer.lr, start_lr)
-
-        callback = LambdaCallback(on_batch_end=lambda batch, logs: self.on_batch_end(batch, logs))
-
-        self.model.fit_generator(generator, epochs=epochs, callbacks=[callback])
-
-        # Restore the weights to the state before model fitting
-        self.model.load_weights('tmp.h5')
+    def find(self, x_train, y_train, start_lr, end_lr, batch_size=64, epochs=1):
+        num_batches = epochs * x_train.shape[0] / batch_size
 
-        # Restore the original learning rate
-        K.set_value(self.model.optimizer.lr, original_lr)
+        self.run(start_lr, end_lr, num_batches, lambda model, callback: (
+            model.fit(x_train, y_train, batch_size=batch_size,
+                        epochs=epochs, callbacks=[callback])
+        ))
+
+    def find_generator(self, generator, start_lr, end_lr, steps_per_epoch=None, epochs=1):
+        num_batches = (
+            epochs * len(generator) if steps_per_epoch == None 
```

Comment:
I would recommend not to do that. 
1. generator can be infinite, or expensive to go through otherwise (e.g. if it loads every batch of data from disk)
2. generator may be an iterable that can be iterated only once. In that case model.fit_generator will fail
I suggest to make steps_per_epoch a required parameter and remove len(generator)

Reply:
I tried to make `find_generator` method to mirror keras's `fit_generator`, just to be consistent and people can call this method exactly the same way they call `fit_generator`.

I'll take a look at keras's source code later on to see how they have dealt with different scenarios.

Reply:
Sounds good.
I think Keras doesn't try to evaluate the length of the generator in fit_generator. It just requires the user to pass steps_per_epoch and uses that as the length of the epoch.
The equivalent code here would be to do this on line 72:
`num_batches = epochs * steps_per_epoch`
and remove default value for steps_per_epoch parameter.